app/uk/gov/hmrc/selfservicetimetopay/controllers/CalculatorController.scala

```diff
@@ -18,94 +18,302 @@ package uk.gov.hmrc.selfservicetimetopay.controllers
+      case Some(json) => Option(JacksonMapper.readValue(json, classOf[CalculatorDuration]))
+      case None => None
+    }
+    val schedules:Option[Seq[CalculatorPaymentSchedule]] = request.session.get("CalculatorPaymentSchedules") match {
+      case Some(json) => Some(generatePaymentSchedules(amountsDue.get.total, paymentToday.get.amount))
+      /*Option(JacksonMapper.readValue(json, classOf[Seq[CalculatorPaymentSchedule]]))*/
+      case None => None
+    }
+    (ssttpStart, amountsDue, paymentToday, duration, schedules)
+  }
+
+  private def paymentScheduleMatches(paymentSchedule: CalculatorPaymentSchedule, amountsDue:CalculatorAmountsDue, paymentToday:CalculatorPaymentToday):Boolean = {
+    (paymentSchedule.amountToPay.compare(amountsDue.total) == 0) && (paymentSchedule.initialPayment.compare(paymentToday.amount.get) == 0)
+  }
+
+  private def getRedirectionDestination(keystoreData:(Boolean, Option[CalculatorAmountsDue], Option[CalculatorPaymentToday], Option[CalculatorDuration], Option[Seq[CalculatorPaymentSchedule]])): Result = {
+    keystoreData match {
+      case (false, _, _, _, _) => {
+        Redirect(routes.SelfServiceTimeToPayController.present)
+      }
+      case (true, None, _, _, _) => {
+        Redirect(routes.CalculatorController.amountsDuePresent())
+      }
+      case (true, a:Some[CalculatorAmountsDue], None, _, _) => {
+        Redirect(routes.CalculatorController.paymentTodayPresent())
+      }
+      case (true, a:Some[CalculatorAmountsDue], p:Some[CalculatorPaymentToday], _, _) => {
+        Redirect(routes.CalculatorController.calculateInstalmentsPresent(None))
+      }
+      case theRest => Redirect(routes.CalculatorController.amountsDuePresent())
+    }
+  }
 
-  def present: Action[AnyContent] = Action.async { implicit request =>
-    Future.successful(Redirect(routes.CalculatorController.amountsDuePresent()).withSession(
-      "CalculatorDuration" -> paymentSchedules.map(_.instalments.length).min.toString
-    ))
+  def present: Action[AnyContent] = Action.async { request =>
+    Future.successful(Redirect(routes.CalculatorController.amountsDuePresent()))
   }
 
   def amountsDuePresent: Action[AnyContent] = Action.async { implicit request =>
-    val form = createAmountDueForm
-    Future.successful(Ok(amounts_due_form.render(formAmountsDue, form, request)))
+    Future.successful(getKeystoreData match {
+
+      case (_, amountsDue:Option[CalculatorAmountsDue], _, _, _) => {
+        Ok(amounts_due_form.render(amountsDue.getOrElse(CalculatorAmountsDue(Seq.empty)), createAmountDueForm, request))
+      }
+
+      case other => getRedirectionDestination(other)
+    })
   }
 
   def amountsDueSubmit: Action[AnyContent] = Action.async { implicit request =>
-    val form = createAmountDueForm.bindFromRequest()
-    if (form.hasErrors) {
-      Future.successful(Ok(amounts_due_form.render(formAmountsDue, form, request)))
-    } else if (form.get.amount.compare(BigDecimal("32.00")) < 0) {
```

Comment:
In fact, I could be wrong but I would have thought that there really is no need to use BigDecimal in the front end, presumably you'll be doing no calculations other than simply adding, if this is the case then a double will suffice, and that will make you code more terse.